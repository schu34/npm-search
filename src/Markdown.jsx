import React, { useState, useEffect } from "react";
import { Box, useInput, Text } from "ink";

const Markdown = ({ source, width, height }) => {
  const [formattedSource, setFormattedSource] = useState([]);
  const [top, setTop] = useState(1);

  useEffect(() => {
    const splitSource = source.split("\n");
    const lines = splitSource.reduce((result, nextLine, lineNumber) => {
      const linesToAppend = [];
      const words = nextLine.split(" ");
      let currentLine = "";
      for (let i = 0; i < words.length; i++) {
        const word = words[i];

        if (currentLine.length + word.length + 1 <= width) {
          currentLine += ` ${word}`;
        } else {
          /* TODO: currently assuming that the window won't be
           * so narrow that one word is too wide for a line.
           * this should be handled better though (maybe we can hyphenate?)
           */
          linesToAppend.push(
            <Text key={currentLine + i + lineNumber}>{currentLine}</Text>
          );
          currentLine = word;
        }
      }
      if (currentLine.length)
        linesToAppend.push(
          <Text key={currentLine + "last" + lineNumber}>{currentLine}</Text>
        );

      return result.concat(linesToAppend);
    }, []);
    setFormattedSource(lines);
  }, [source]);

  useInput((input, key) => {
    if (key.upArrow || input === "k") {
      return setTop(Math.max(top - 1, 0));
    }

    if (key.downArrow || input === "j") {
      return setTop(Math.min(formattedSource.length - height, top + 1));
    }
  });

  return (
    <Box flexDirection={"column"} width={width} height={height}>
      {formattedSource.slice(top, top + height - 1)}
    </Box>
  );
};

export default Markdown;
