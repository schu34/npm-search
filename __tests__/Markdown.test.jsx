import React from "react";
import fs from "fs";
import Markdown from "../src/Markdown";
import { render } from "ink-testing-library";

const mdString = fs.readFileSync("./readme.md").toString();

describe("<Markdown />", () => {
  it("should render without errors", () => {
    const { lastFrame, rerender } = render(
      <Markdown source={mdString} width={50} height={100} />
    );
    rerender(<Markdown source={mdString} width={50} height={100} />);
    expect(lastFrame()).toMatchSnapshot();
  });
});
