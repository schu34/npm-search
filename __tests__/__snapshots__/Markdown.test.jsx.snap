// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Markdown /> should render without errors 1`] = `
" # npm-search

 npm-search allows you to search for and install
npm packages, all from the comfort of you
terminal.

 ## installation

 Not published yet, I'll add instructions here
once it is

 ## usage

 Just \`npm-search\` in your terminal, the app will
guide you through the rest

 ## TODOs

 1. ~allow users to search~
 1. ~Refactor routing to use less ternaries~
 1. ~add details view which allows users to read
package READMEs~
 1. infinite scroll for package search results
 1. allow users to install packages from within
the app using \`i\` from the details screen
 1. add instructions display as bottom line(tells
users which keys to push to use functionality)
 1. allow user to start a new search from any page
in the app.
 1. add command line argument support for \`-s\`
option. users should be able to do \`npm-search
-s=express\` to go directly to the search results
for \`express\`
 1. add command line argument support for the \`-p\`
option. users should be able to do \`npm-search
-p=express\` to go directly to the details page for
\`express\`






























































"
`;
